# Lab 1 - Enter the world of Patents

In this lab, we will explore the world of patents. We will use the data from 
European Patent Office (EPO) to extract information about patents. In this lab,
you will learn:

- Basic structure of a patent
- Querying patents using the EPO API via SPARQL (a query language for RDF)
- Extracting information from patents
- Visualizing the extracted information
- Summarizing the extracted information

I am very excited to start this journey with you as we will disover many interesting
things about patents. Let's get started!

In [2]:
# install the package
install.packages("../_utils/SPARQL_1.16.tar.gz", repos = NULL, type="source")

Installing package into ‘/home/michael/R/x86_64-pc-linux-gnu-library/4.3’
(as ‘lib’ is unspecified)



In [3]:
library(pacman)
p_load(readxl, stringr, data.table, magrittr, ggplot2, SPARQL,
        eurostat, XML, RCurl, knitr)

# color palette
gray_scale <- c('#F3F4F8','#D2D4DA', '#B3B5BD', 
                '#9496A1', '#7d7f89', '#777986', 
                '#656673', '#5B5D6B', '#4d505e',
                '#404352', '#2b2d3b', '#282A3A',
                '#1b1c2a', '#191a2b',
                '#141626', '#101223')

ft_palette <- c('#990F3D', '#0D7680', '#0F5499', '#262A33', '#FFF1E5')

ft_contrast <- c('#F83', '#00A0DD', '#C00', '#006F9B', '#F2DFCE', '#FF7FAA',
                 '#00994D', '#593380')

peep_head <- function(dt, n = 5) {
    dt %>%
        head(n) %>%
        kable()
}

peep_sample <- function(dt, n = 5) {
    dt %>%
        .[sample(.N, n)] %>%
        kable()
}

peep_tail <- function(dt, n = 5) {
    dt %>%
        tail(n) %>%
        kable()
}

## Patents that mention specific words in the abstract

In [10]:
# query the SPARQL endpoint
endpoint <- "https://data.epo.org/linked-data/query"

# patents that abstract contains the word 'battery'
query1 <- "
prefix dcterms: <http://purl.org/dc/terms/>
prefix patent: <http://data.epo.org/linked-data/def/patent/>
prefix text: <http://jena.apache.org/text#>

SELECT DISTINCT ?publication ?title ?abstract 
WHERE {
    ?publication text:query ( dcterms:abstract 'battery' );
                 patent:titleOfInvention ?title;
                 dcterms:abstract        ?abstract.
} LIMIT 10
"
query1_result <- SPARQL(endpoint, query1)$results

In [12]:
str(query1_result)

'data.frame':	10 obs. of  3 variables:
 $ publication: chr  "<http://data.epo.org/linked-data/data/publication/EP/1182716/A3/->" "<http://data.epo.org/linked-data/data/publication/EP/1182716/A3/->" "<http://data.epo.org/linked-data/data/publication/EP/1182716/A3/->" "<http://data.epo.org/linked-data/data/publication/EP/1182716/A2/->" ...
 $ title      : chr  "\"Batteriehalterung\"@de" "\"Support de batterie\"@fr" "\"Battery lock\"@en" "\"Batteriehalterung\"@de" ...
 $ abstract   : chr  "\"A battery lock for a communication unit with holding means for holding an internal battery in operating posit"| __truncated__ "\"A battery lock for a communication unit with holding means for holding an internal battery in operating posit"| __truncated__ "\"A battery lock for a communication unit with holding means for holding an internal battery in operating posit"| __truncated__ "\"A battery lock for a communication unit with holding means for holding an internal battery in operating posit"| __trunc

In [11]:
head(query1_result)

publication                                                       
1 <http://data.epo.org/linked-data/data/publication/EP/1182716/A3/->
2 <http://data.epo.org/linked-data/data/publication/EP/1182716/A3/->
3 <http://data.epo.org/linked-data/data/publication/EP/1182716/A3/->
4 <http://data.epo.org/linked-data/data/publication/EP/1182716/A2/->
5 <http://data.epo.org/linked-data/data/publication/EP/1182716/A2/->
6 <http://data.epo.org/linked-data/data/publication/EP/1182716/A2/->
  title                   
1 "Batteriehalterung"@de  
2 "Support de batterie"@fr
3 "Battery lock"@en       
4 "Batteriehalterung"@de  
5 "Support de batterie"@fr
6 "Battery lock"@en       
  abstract                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           
1 "A battery lock for a communication unit with holding means for holding an internal battery in operating position in a battery compartment. The holding means includes features to hold, lift and release the battery in the compartment from the battery compartment. The battery compartment includes protruding parts corresponding to slots on the battery, where the protruding parts co-operates with holding means of the battery to hold the battery in the battery compartment. The holding means of the battery lock includes first protruding parts to hold the battery in the battery compartment, grips to lift and release the battery, holes to hinge the battery lock in the battery compartment, second protruding parts to lift the battery actuated by the grips and fastening means to fasten the battery lock to the battery compartment. The fastening means on the battery compartment to fasten the battery lock to the battery compartment include pivots to hinge the battery on, locking flaps to keep the battery lock on the pivots and protruding parts that connects to slots on the grips of the battery lock. "@en
2 "A battery lock for a communication unit with holding means for holding an internal battery in operating position in a battery compartment. The holding means includes features to hold, lift and release the battery in the compartment from the battery compartment. The battery compartment includes protruding parts corresponding to slots on the battery, where the protruding parts co-operates with holding means of the battery to hold the battery in the battery compartment. The holding means of the battery lock includes first protruding parts to hold the battery in the battery compartment, grips to lift and release the battery, holes to hinge the battery lock in the battery compartment, second protruding parts to lift the battery actuated by the grips and fastening means to fasten the battery lock to the battery compartment. The fastening means on the battery compartment to fasten the battery lock to the battery compartment include pivots to hinge the battery on, locking flaps to keep the battery lock on the pivots and protruding parts that connects to slots on the grips of the battery lock. "@en


In [13]:
# make it more readable
query1_result %>%
    as.data.table() %>%
    # delete '<' and '>' for publication
    .[, publication := gsub("[<>]", "", publication)] %>%
    peep_head()



|publication                                                      |title                    |abstract                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 

Please **click the link in the above table** to check the full information about the patents. Notice that we have same patent number - **1182716** - in the list. This is because the same patent but different publications are listed in the results.

Here is some basic information about the patents:

- A document: European patent application, published 18 months after filing with the EPO or 18 months after priority date. 

    - A1 document: European patent application published with European search report
    - A2 document: European patent application published without European search report (search report not available at the publication date)
    - A3 document: Separate publication of the European search report
    - A4 document: Supplementary search report

- B document:European patent specification

    - B1 document: European patent specification (granted patent)
    - B2 document: New European patent specification (amended specification)
    - B3 document: European patent specification (after limitation procedure)


> Anyone can apply for a patent, but not every patent will be granted!

![patent-application-process](../images/IPRIS_The-Process-Diagram_01_0.png)